src/coord/src/coord.rs

```diff
@@ -1484,6 +1334,175 @@ impl Coordinator {
+                    Statement::AlterIndex(_)
+                    | Statement::AlterObjectRename(_)
+                    | Statement::CreateDatabase(_)
+                    | Statement::CreateIndex(_)
+                    | Statement::CreateRole(_)
+                    | Statement::CreateCluster(_)
+                    | Statement::CreateSchema(_)
+                    | Statement::CreateSecret(_)
+                    | Statement::CreateSink(_)
+                    | Statement::CreateSource(_)
+                    | Statement::CreateTable(_)
+                    | Statement::CreateType(_)
+                    | Statement::CreateView(_)
+                    | Statement::CreateViews(_)
+                    | Statement::Delete(_)
+                    | Statement::DropDatabase(_)
+                    | Statement::DropObjects(_)
+                    | Statement::Insert(_)
+                    | Statement::Update(_) => {
+                        return tx.send(
+                            Err(CoordError::OperationProhibitsTransaction(stmt.to_string())),
+                            session,
+                        )
+                    }
+                }
+            }
+        }
+
+        if self.catalog.config().safe_mode {
+            if let Err(e) = check_statement_safety(&stmt) {
+                return tx.send(Err(e), session);
+            }
+        }
+
+        let stmt = stmt.clone();
+        let params = portal.parameters.clone();
+        match stmt {
+            // `CREATE SOURCE` statements must be purified off the main
+            // coordinator thread of control.
+            Statement::CreateSource(stmt) => {
+                let internal_cmd_tx = self.internal_cmd_tx.clone();
+                let conn_id = session.conn_id();
+                let params = portal.parameters.clone();
+                let purify_fut = mz_sql::pure::purify_create_source(
+                    self.now(),
+                    self.catalog.config().aws_external_id.clone(),
+                    stmt,
+                );
+                task::spawn(|| format!("purify:{conn_id}"), async move {
+                    let result = purify_fut.err_into().await;
+                    internal_cmd_tx
+                        .send(Message::CreateSourceStatementReady(
+                            CreateSourceStatementReady {
+                                session,
+                                tx,
+                                result,
+                                params,
+                            },
+                        ))
+                        .expect("sending to internal_cmd_tx cannot fail");
+                });
+            }
+
+            // All other statements are handled immediately.
+            _ => match self.handle_statement(&mut session, stmt, &params).await {
+                Ok(plan) => self.sequence_plan(tx, session, plan).await,
+                Err(e) => tx.send(Err(e), session),
+            },
+        }
+    }
```

Comment:
is this whole function copied? or were there any changes (other than the purify_create_source)?

Reply:
Yup, copied nearly verbatim. The only difference is that `tx` is now a `ClientTransmitter` instead of a `oneshot::Sender` directly, which just amounmts to all of the error handling taking one line instead of four.